server/pulp/plugins/distributor.py

```diff
@@ -337,3 +355,20 @@ def cancel_publish_group(self, call_request, call_report):
         :raise SystemExit: raised through a call to sys.exit()
         """
         sys.exit()
+
+    @staticmethod
+    def ensure_all_units_downloaded(repo_group):
+        """
+        Checks the database to make sure all units in the repos have been downloaded. If not, raises
+        an exception.
+
+        :param repo_group:  RepositoryGroup instance that should be checked
+        :type  repo_group:  pulp.plugins.model.RepositoryGroup
+
+        :raises exceptions.PulpCodedException:  if any unit in the group has un-downloaded files
+        """
+        bad_repo_ids = filter(lambda x: not repo_controller.has_all_units_downloaded(x),
+                              repo_group.repo_ids or [])
```

Comment:
I looked at the model and a bit of how it's used, and I couldn't reach a confident conclusion that it will always be a list. Beyond current code, we have all the code from the past that could have written this document with a repo_ids attribute of ``None``.

It's unlikely I admit, but we've been bit by that kind of bug many times in the past. I don't think there's anything incorrect that could come from the ``or []``, but it does give us a line of defense in case such a repo group makes its way through this code.

So I concluded it's better to take the defensive option, especially since it's so easy, than risk even a small chance of a user or customer getting a traceback here.